drivers/flash/nrf_qspi_nor.c

```diff
@@ -0,0 +1,673 @@
+ *
+ * @param buf is a valid pointer to a data buffer.
+ * Can not be NULL.
+ * @param len is the length of the data to be handled.
+ * If no data to transmit/receive - pass 0.
+ */
+struct qspi_buf {
+	u8_t *buf;
+	size_t len;
+};
+
+/**
+ * @brief QSPI command structure
+ * Structure used for custom command usage.
+ *
+ * @param op_code is a command value (i.e 0x9F - get Jedec ID)
+ * @param tx_buf structure used for TX purposes. Can be NULL if not used.
+ * @param rx_buf structure used for RX purposes. Can be NULL if not used.
+ */
+struct qspi_cmd {
+	u8_t op_code;
+	const struct qspi_buf *tx_buf;
+	const struct qspi_buf *rx_buf;
+};
+
+/**
+ * @brief Structure for defining the QSPI NOR access
+ * @param sem The semaphore to access to the flash
+ * @param sync The semaphore to ensure that transfer has finished
+ * @param write_protection Indicates if write protection for flash
+ *  device is enabled
+ */
+struct qspi_nor_data {
+	struct k_sem sem;
+	struct k_sem sync;
+	bool write_protection;
+};
+
+static inline int qspi_get_mode(bool cpol, bool cpha)
+{
+	register int ret = -EINVAL;
+
+	if ((!cpol) && (!cpha)) {
+		ret = 0;
+	} else if (cpol && cpha) {
+		ret = 1;
+	}
+	__ASSERT(ret != -EINVAL, "Invalid QSPI mode");
+	return ret;
+}
+
+static inline bool qspi_is_used_write_quad_mode(nrf_qspi_writeoc_t lines)
+{
+	switch (lines) {
+	case NRF_QSPI_WRITEOC_PP4IO:
+	case NRF_QSPI_WRITEOC_PP4O:
+		return true;
+	default:
+		return false;
+	}
+}
+
+static inline bool qspi_is_used_read_quad_mode(nrf_qspi_readoc_t lines)
+{
+	switch (lines) {
+	case NRF_QSPI_READOC_READ4IO:
+	case NRF_QSPI_READOC_READ4O:
+		return true;
+	default:
+		return false;
+	}
+}
+
+static inline int qspi_get_lines_write(u8_t lines)
+{
+	register int ret = -EINVAL;
+
+	switch (lines) {
+	case 3:
+		ret =  NRF_QSPI_WRITEOC_PP4IO;
+		break;
+	case 2:
+		ret = NRF_QSPI_WRITEOC_PP4O;
+		break;
+	case 1:
+		ret = NRF_QSPI_WRITEOC_PP2O;
+		break;
+	case 0:
+		ret = NRF_QSPI_WRITEOC_PP;
+		break;
+	default:
+		break;
+	}
+	__ASSERT(ret != -EINVAL, "Invalid QSPI write line");
+	return ret;
+}
+
+static inline int qspi_get_lines_read(u8_t lines)
+{
+	register int ret = -EINVAL;
+
+	switch (lines) {
+	case 4:
+		ret = NRF_QSPI_READOC_READ4IO;
+		break;
+	case 3:
+		ret = NRF_QSPI_READOC_READ4O;
+		break;
+	case 2:
+		ret = NRF_QSPI_READOC_READ2IO;
+		break;
+	case 1:
+		ret = NRF_QSPI_READOC_READ2O;
+		break;
+	case 0:
+		ret = NRF_QSPI_READOC_FASTREAD;
+		break;
+	default:
+		break;
+	}
+	__ASSERT(ret != -EINVAL, "Invalid QSPI read line");
+	return ret;
+}
+
+/**
+ * @brief Get QSPI prescaler
+ * Get supported frequency prescaler not exceeding the requested one.
+ *
+ * @param frequency - desired QSPI bus frequency
+ * @retval NRF_SPI_PRESCALER in case of success or;
+ *		   -EINVAL in case of failure
+ */
+static inline nrf_qspi_frequency_t get_nrf_qspi_prescaler(u32_t frequency)
+{
+	register int ret = -EINVAL;
+
+	if (frequency > 2000000) {
```

Comment:
this comparison stages look erroneous as ensure else case in second compassion.

Reply:
just 2000000ul

Reply:
IMO this should be `2000000UL`, i.e. upper case.  We don't want a lower case `l` because it looks too much like `1`, so for consistency it should be `UL`.  Also see line 172 where `32000000uL` looks weird.

Not a blocking comment, just a preference.

Reply:
I have added changes due to yours request.